webtasks/cron_mongodb.js

```diff
@@ -0,0 +1,230 @@
+            });
+    } else if (action === 'reserve_jobs') {
+        // Reserve a set of jobs for execution
+        
+        if (!validate_params(['count', 'ttl']))
+            return;
+        
+        var now = new Date();
+        var next_available_at = new Date(now.valueOf() + (parseInt(context.data.ttl, 10) * 1000))
+        
+        return withMongoCollection(context.data.MONGO_COLLECTION)
+            .then(function (coll) {
+                return Bluebird.resolve(_.range(context.data.count))
+                    // TODO (ggoodman): Refactor this to use reduce to avoid calling N queries if reservation fails before N
+                    .map(function (n) {
+                        console.log("Attempting reservation", n);
+                        var filter = {
+                            next_available_at: {
+                                $lte: now
+                            },
+                            state: 'active',
+                        };
+                        var update = {
+                            $set: {
+                                next_available_at: next_available_at
+                            }
+                        };
+                        
+                        return coll.findOneAndUpdateAsync(filter, update, {
+                            returnOriginal: false, // Return modified
+                        })
+                            .get('value'); // Only pull out the value
+                        
+                    }, {concurrency: 1}) // Serial mapping
+                    .filter(Boolean);
+                
+            })
+            .then(function (data) {
+                res.writeHead(200, {'Content-Type': 'application/json'});
+                res.end(JSON.stringify(data));
+            })
+            .catch(function (err) {
+                error(err.statusCode || 500, err);
+            });
+    } else if (action === 'list_jobs') {
+        // List all scheduled webtasks or only those in a container passed via query
+        
+        return withMongoCollection(context.data.MONGO_COLLECTION)
+            .then(function (coll) {
+                var query = {};
+                
+                // Both admin and user will hit this endpoint. When admin, container can be unset
+                if (context.data.container) query.container = context.data.container;
+                
+                // TODO (ggoodman): Pagination logic
+                return coll.findAsync(query)
+                    .then(function (cursor) {
+                        return Bluebird.promisify(cursor.toArray, cursor)();
+                    });
+            })
+            .then(function (data) {
+                res.writeHead(200, {'Content-Type': 'application/json'});
+                res.end(JSON.stringify(data));
+            })
+            .catch(function (err) {
+                error(err.statusCode || 500, err);
+            });
+    
+    } else if (action === 'get_job') {
```

Comment:
Since the former is working on a single item and the latter on an array of items, each time a step needs to be modified to account for the different nature, we introduce a branch. I prefer to keep the branch at the top level. As you will see in the `stripMongoId` commit, this means `.map(stripMongoId)` vs. `.filter(stripMongoId)` and the difference in query operator you mentioned.

If you feel strongly about this, I can parametrize the methods based on the action to reuse common logic.